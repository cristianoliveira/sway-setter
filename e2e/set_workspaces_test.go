package e2e

import (
	"bytes"
	"fmt"
	"os/exec"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
)

func TestSetWorkspaces(t *testing.T) {
	cases := []struct {
		name       string
		args       []string
		stdin      string
		expectFail bool
	}{
		{
			name: "set unkonwn workspaces",
			args: []string{
				"-t", "unknown",
				"--print",
			},
			stdin:      `[{"id":1,"name":"1","output":"HDMI-A-0","focused":true}]`,
			expectFail: true,
		},
		{
			name: "set workspaces with invalid json",
			args: []string{
				"-t", "set_workspaces",
				"--print",
			},
			// Expect a json array
			stdin:      `{"id":1,"name":"1","output":"HDMI-A-0","focused":true}`,
			expectFail: true,
		},
		{
			name: "set workspaces with incomplete json",
			args: []string{
				"-t", "set_workspaces",
				"--print",
			},
			// Expect a json array
			stdin:      `{"id":1,"name":"1"`,
			expectFail: true,
		},
		{
			name: "set workspaces one workspace",
			args: []string{
				"-t", "set_workspaces",
				"--print",
			},
			stdin: `[
				{
					"id":1,
					"name":"1",
					"output":"HDMI-A-0",
					"focused":true
				}
			]`,
			expectFail: false,
		},
		{
			name: "set workspaces with multiple workspaces",
			args: []string{
				"-t", "set_workspaces",
				"--print",
			},
			stdin: `[
				{
					"id":1,
					"name":"1",
					"output":"HDMI-A-0",
					"focused":false
				},
				{
					"id":2,
					"name":"2",
					"output":"DP-1",
					"focused":true
				},
				{
					"id":3,
					"name":"3",
					"output":"eDP-1",
					"focused":false
				}
			]`,
			expectFail: false,
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(tt *testing.T) {
			cmd := exec.Command("../bin/sway-setter", tc.args...)

			var out, errOut bytes.Buffer
			stdin := bytes.NewBufferString(tc.stdin)
			cmd.Stdin = stdin
			cmd.Stdout = &out
			cmd.Stderr = &errOut

			if err := cmd.Run(); err != nil && !tc.expectFail {
				tt.Fatalf("\nUnexpected fail:\n%s\nstderr:%s\n stdout:%s", err, errOut.String(), out.String())
			}

			snaps.MatchSnapshot(
				tt,
				fmt.Sprintf("%s < data.json", cmd.String()),
				out.String(),
			)
		})
	}
}
