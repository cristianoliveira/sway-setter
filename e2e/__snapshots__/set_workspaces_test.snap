
[TestSetWorkspaces/set_unkonwn_workspaces - 1]
../bin/sway-setter -t unknown --print < data.json
Error: type `unknown` is not supported
Supported types
 set_workspaces: load workspaces from `swaymsg` output

---

[TestSetWorkspaces/set_workspaces_with_invalid_json - 1]
../bin/sway-setter -t set_workspaces --print < data.json
Error: failed to parse input content
Hint: make sure to use a file generated by the `swaymsg -t get_workspaces` output

---

[TestSetWorkspaces/set_workspaces_with_incomplete_json - 1]
../bin/sway-setter -t set_workspaces --print < data.json
Error: failed to parse input content
Hint: make sure to use a file generated by the `swaymsg -t get_workspaces` output

---

[TestSetWorkspaces/set_workspaces_one_workspace - 1]
../bin/sway-setter -t set_workspaces --print < data.json
workspace 1; move workspace to output HDMI-A-0
workspace 1

---

[TestSetWorkspaces/set_workspaces_with_multiple_workspaces - 1]
../bin/sway-setter -t set_workspaces --print < data.json
workspace 1; move workspace to output HDMI-A-0
workspace 2; move workspace to output DP-1
workspace 3; move workspace to output eDP-1
workspace 2

---
